cmd/tpm-manager/tpm-manager.go

```diff
@@ -0,0 +1,544 @@
+		err = invalidateNode(node)
+		if err != nil {
+			return fmt.Errorf("Failed to invalidate node: %v", err)
+		}
+		return fmt.Errorf("Invalidating Node: Unable to obtain TPM quote for node %s: %v", address.String(), err)
+	}
+
+	err = tpm.ValidateLog(log, quote)
+	if err != nil {
+		err = invalidateNode(node)
+		if err != nil {
+			return fmt.Errorf("Failed to invalidate node: %v", err)
+		}
+		return fmt.Errorf("Invalidating Node: TPM event log does not match quote for node %s", address.String())
+	}
+
+	// Don't handle this error immediately - we want to update the annotations even if validation failed
+	logstate, err := tpm.ValidatePCRs(log, quote, manager.pcrConfigs)
+	jsonlog, jsonerr := json.Marshal(logstate)
+
+	if jsonerr == nil {
+		updateAnnotations(node, string(jsonlog))
+	} else {
+		glog.Errorf("Failed to serialise new log state: %v", jsonerr)
+	}
+
+	if err != nil {
+		err = invalidateNode(node)
+		if err != nil {
+			return fmt.Errorf("Failed to invalidate node: %v", err)
+		}
+		return fmt.Errorf("Invalidating Node: Unable to validate quote for node %s", address.String())
+	}
+
+	// If we've got this far then the node is trustworthy
+	err = trustNode(node)
+	if err != nil {
+		return fmt.Errorf("Failed to mark node as trustworthy")
+	}
+	return nil
+}
+
+func verifyAndUpdate(node *api.Node) {
+	err := verifyNode(node)
+	if err != nil {
+		glog.Errorf("Failed to verify node %s: %v", node.Name, err)
+	}
+	// The state that the node will be updated to
+	newstate := isTrusted(node)
+	newnode, err := manager.client.Nodes().Get(node.Name)
+	if err != nil {
+		glog.Errorf("Unable to obtain node state for %s: %v", node.Name, err)
+		return
+	}
+	if newnode.ObjectMeta.Annotations == nil {
+		newnode.ObjectMeta.Annotations = make(map[string]string)
+	}
+	currenttime := time.Now().Unix()
+
+	// If we're transitioning state, update the metadata
+	if isTrusted(newnode) != newstate {
+		if newstate == false {
+			newnode.ObjectMeta.Annotations[UntrustedSince] = strconv.FormatInt(currenttime, 10)
+			newnode.ObjectMeta.Annotations[TrustedSince] = ""
+		} else {
+			newnode.ObjectMeta.Annotations[UntrustedSince] = ""
+			newnode.ObjectMeta.Annotations[TrustedSince] = strconv.FormatInt(currenttime, 10)
+		}
+	}
+
+	newnode.ObjectMeta.Annotations[ValidationTime] = strconv.FormatInt(currenttime, 10)
+
+	// Ensure that the new node is tainted appropriately
+	if newstate == true {
+		trustNode(newnode)
+	} else {
+		invalidateNode(newnode)
+	}
+	newnode.ObjectMeta.Annotations[LogState] = node.ObjectMeta.Annotations[LogState]
+	newnode, err = manager.client.Nodes().Update(newnode)
+	if err != nil {
+		glog.Errorf("Unable to update node state for %s: %v", node.Name, err)
+		return
+	}
+}
+
+func verifyAllNodes() {
+	nodes, err := manager.client.Nodes().List(api.ListOptions{})
+	if err != nil {
+		glog.Errorf("Unable to obtain list of nodes")
+		return
+	}
+	for _, node := range nodes.Items {
+		verifyAndUpdate(&node)
+	}
+}
+
+func reverify() {
```

Comment:
I'm not clear on what you mean about changing watch at runtime. What configMap? What about watch needs to change?

More or less what I was (psuedo) thinking:

Add a node store to the TPMManager, then make use of the resyncPeriod/updateFunc of the informer to do periodic re-verification, then also use the nodeStore cache to do a node listing, rather than doing full round-trip to api-server each time (we get the fresh node record in verifyAndUpdate anyway):

```
func run() {
    nodeStore, nodeController := framework.NewInformer(
        ...
        updateFunc: verifyAllNodes() // fires on resync
        resyncPeriod: manager.recurring * time.Second
    )
    mananger.nodeStore = nodeStore
    go nodeController.Run()
}

func verifyAllNodes() {
    nodes, err := manager.nodeStore.List() // Retrieve from local cache
    for _, node := range nodes.Items {
        verifyAndUpdate(&node) // We retrieve fresh node object anway from api
    }
}
```

But this is just brainstorm -- may not work as you need